xarray/core/dataset.py

```diff
@@ -1933,6 +1933,47 @@ def from_dataframe(cls, dataframe):
             obj[name] = (dims, data)
         return obj
 
+    def to_dict(self):
+        """Convert this dataset to a dictionary following xarray naming conventions.
+        """
+        d = {'coords': {}, 'attrs': dict(self.attrs), 'dims': list(self.dims), 
+             'data_vars': {}}
+
+        for k in self.coords:
+            d['coords'].update({k: {'data': self[k].values.tolist(),
+                                    'dims': list(self[k].dims),
+                                    'attrs': dict(self[k].attrs)}})
+        for k in self.data_vars:
+            d['data_vars'].update({k: {'data': self[k].values.tolist(),
+                                       'dims': list(self[k].dims),
+                                       'attrs': dict(self[k].attrs)}})
+        return d
+
+    @classmethod
+    def from_dict(cls, d):
+        """
+        Convert a dictionary into an xarray.Dataset.
+        """
+        obj = cls()
```

Comment:
Because of the way that Dataset works internally, it can be much slower (for a large number of variables) to add them incrementally to a Dataset rather than to call the Dataset constructor once. Also, the behavior could differ, because the variables are aligned incrementally rather than all at once.

So I would consider building up ordered dictionaries for `data_vars`, `coords` and `attrs` and calling `cls(data_vars, coords, attrs)` at the end.